File: swift-backend/Sources/App/Services/SyncService.swift
@preconcurrency import Fluent
import Plaid
import Vapor

struct SyncService {
    let plaid: PlaidClient
    let plaidItemService: PlaidItemService
    let accountService: AccountService
    let transactionService: TransactionService

    // todo: move task to a persistent queue
    func syncTransactionsAndAccounts(itemID: UUID, logger: Logger) {
        Task {
            do {
                guard let item = try await plaidItemService.getByID(id: itemID) else {
                    logger.error(
                        "Cannot sync transactions. Failed to fetch item with id \(itemID)"
                    )
                    return
                }

                let batchSize = 100
                var cursor = item.transactionsCursor
                var hasMore = true
                var added: [Components.Schemas.Transaction] = []
                var updated: [Components.Schemas.Transaction] = []
                var removed: [Components.Schemas.RemovedTransaction] = []

                while hasMore {
                    let data = try await plaid.getTransactionsSync(
                        plaidItemID: item.plaidItemID,
                        accessToken: item.plaidAccessToken,
                        cursor: cursor,
                        count: batchSize
                    )

                    added.append(contentsOf: data.added)
                    updated.append(contentsOf: data.modified)
                    removed.append(contentsOf: data.removed)

                    hasMore = data.has_more
                    cursor = data.next_cursor
                }

                let accountsResponse = try await plaid.getAccounts(
                    accessToken: item.plaidAccessToken
                )
                try await accountService.upsertAccounts(
                    plaidItemID: item.plaidItemID,
                    accounts: accountsResponse.accounts
                )
                try await transactionService.upsertTransactions(
                    transactions: added + updated
                )
                try await transactionService.deleteTransactions(transactions: removed)
                guard let cursor = cursor else {
                    logger.error("Cursor is nil. Cannot update cursor.")
                    return
                }
                try await plaidItemService.updateCursor(
                    itemID: try item.requireID(),
                    cursor: cursor
                )

            }
            catch { print("Failed to sync plaid: \(String(reflecting: error))") }
        }
    }
}

File: ts-backend/src/sync_service.ts
import { Brand } from "effect";
import { Db } from "./db";
import { Transaction, RemovedTransaction, PlaidApi } from "plaid";

type ItemId = string & Brand.Brand<"ItemId">;

export const ItemId = Brand.nominal<ItemId>();

export class SyncService {
  static async syncAccountsAndTransactions(
    db: Db,
    plaidClient: PlaidApi,
    itemId: ItemId,
  ): Promise<void> {
    const item = await db.query.plaidItems.findFirst({
      where: (items, { eq }) => eq(items.id, itemId.toString()),
    });

    if (!item) {
      throw new Error(
        `Cannot sync transactions. Failed to fetch item with id $itemID`,
      );
    }

    let batchSize = 100;
    let cursor = item.transactions_cursor ?? undefined;
    let hasMore = true;
    const addedTransactions: Transaction[] = [];
    const modifiedTransactions: Transaction[] = [];
    const removedTransactions: RemovedTransaction[] = [];

    while (hasMore) {
      const response = await plaidClient.transactionsSync({
        access_token: item.plaid_access_token,
        count: batchSize,
        cursor,
      });

      addedTransactions.push(...response.data.added);
      modifiedTransactions.push(...response.data.modified);
      removedTransactions.push(...response.data.removed);

      hasMore = response.data.has_more;
      cursor = response.data.next_cursor;
    }

    console.log({
      addedTransactions: addedTransactions.length,
      modifiedTransactions: modifiedTransactions.length,
      removedTransactions: removedTransactions.length,
    });

    // get all accounts
    // upsert accounts
    // upsert transactions
    // delete transactions
    // update item with new cursor
  }
}
